text/projects/Grzegorz_Gawryał/Matching_with_dont_cares_and_mismatches.tex

```diff
@@ -0,0 +1,149 @@
+\date{Maj 2022}
+
+\begin{document}
+
+\maketitle
+
+\section{Wstęp}
+W problemie wyszukiwania wzorca z wieloznacznikami i kilkoma błędami dany jest tekst $t$ długości $n$ oraz wzorzec $w$, długości $m$, składające się ze słów nad alfabetem $\Sigma \cup \{?\}$. Dla każdej pozycji $i\in [n-m+1]$ chcemy stwierdzić, czy odległość Hamminga między słowami $t[i\dots i+m-1]$ oraz $w$ wynosi co najwyżej $k$, przy czym symbol specjalny $?$ pasuje do każdego innego symbolu (w szczególności do $?$). Praca korzysta z obserwacji poczynionych w pracy ''Simple deterministic wildcard matching'' P. Clifforda i R. Clifforda, podając główny algorytm randomizowany, a następnie modyfikując go na dwa sposoby, przyśpieszając go sprytniej wykorzystując dotychczasowe informacje oraz derandomizując go, stosując wyniki uzyskane dla znanego problemu z teorii informacji.
+
+\section{Algorytmy i dowody}
+\subsection{Wynik pracy ''Simple deterministic wildcard matching'' P. Clifforda i R. Clifforda}
+W tej pracy autorzy rozważali problem  wyszukiwania wzorca z wieloznacznikami i bez błędów ($k=0$). Niech $f: \Sigma \cup \{?\} \to \{0,\dots, |\Sigma|\}$ będzie dowolną funkcją różnowartościową taką, że $f(?) = 0$.
+Autorzy zauważyli, że $w$ pasuje do podsłowa $t[i\dots i+m-1]$ wtw, gdy 
+\[A_0(i) = \sum_{j=1}^m f(w_j) f(t_{i+j-1})(f(w_j) - f(t_{i+j-1}))^2 = 0\]
+Można zauważyć, że ta suma będzie zerowa wtw, gdy każdy jej składnik będzie zerowy, czyli $f(w_j) = 0,\ f(t_{i+j-1}) = 0$, więc któryś z tych symboli jest wieloznacznikiem, lub $f(w_j) = f(t_{i+j-1})$, czyli symbole do siebie pasują.
+
+Powyższą sumę dla każdego $i$ można obliczyć poprzez klasyczny splot $(+, *)$, wykonując go na ciągach:
+\begin{itemize}
+    \item $\Big(f(w_1)^3, \dots, f(w_m)^3\Big)$ oraz $\Big(f(t_{i+m-1}), \dots, f(t_{i})\Big)$
+    \item $\Big(-2f(w_1)^2, \dots, -2f(w_m)^3\Big)$ oraz $\Big(f(t_{i+m-1})^2, \dots, f(t_{i})^2\Big)$
+    \item $\Big(f(w_1), \dots, f(w_m)\Big)$ oraz $\Big(f(t_{i+m-1})^3, \dots, f(t_{i})^3\Big)$
+\end{itemize}
+a następnie sumując otrzymane wektory. Całość zajmuje czas $O(n \log m)$, wykorzystując FFT.
+
+\subsection{Algorytm dla $k\le1$}
+Lekko zmodyfikujmy poprzednią sumę:
+\[A_1(i) = \sum_{j=1}^m (i+j-1)f(w_j) f(t_{i+j-1})(f(w_j) - f(t_{i+j-1}))^2\]
+Zauważmy, że jeśli $w$ różni się na dokładnie jednej pozycji $x$ od podsłowa $t[i\dots i+m-1]$, to $A_0(i) = f(w_x) f(t_{i+x-1})(f(w_x) - f(t_{i+x-1}))^2$ oraz $A_1(i) = (i+x-1)f(w_x) f(t_{i+x-1})(f(w_x) - f(t_{i+x-1}))^2$. Dla każdego indeksu $i$ zachodzi więc dokładnie jeden z przypadków:
+\begin{itemize}
+    \item $A_0(i) = 0$ -- wtedy odpowiednie słowa pasują do siebie bez błędów,
+    \item $A_0(i) \ne 0$ -- wówczas liczymy $B(i) = \frac{A_1(i)}{A_0(i)}$. Jeśli odpowiednie słowa różnią się na dokładnie jednej pozycji, to będzie to indeks $y$. Sprawdzamy więc, czy $A_0(i) = (i+y-1)f(w_y) f(t_{i+y-1})(f(w_y) - f(t_{i+y-1}))^2$ -- jeśli tak, to będzie to jedyny błąd, a jeśli nie, to znaczy, że mamy więcej błędów.
```

Comment:
"Sprawdzamy więc, czy $A_0(i) = (i+y-1)f(w_y) f(t_{i+y-1})(f(w_y) - f(t_{i+y-1}))^2$" - shouldn't we check rather this way: compute $y = B(i)$ and see whether $t[i + y - 1] \neq w[y]$? I don't see why your check works.

Reply:
There was a typo, (i+y-1) should be omitted. I think both versions  $t_{i+y-1} \ne w_y$ and $A_0(i) = f(w_y) f(t_{i+y-1})(f(w_y) - f(t_{i+y-1}))^2$ are correct, because either:
* $f(w_y)$ and $f(t_{i+y-1})$ is the only mismatch, then both these predicates are true
* $f(w_y)$ and $f(t_{i+y-1})$ is a mismatch, but not the only one - it's not possible, because then $y = B(i)$ would be greater, possibly non integer, but since algorithm uses integer division, it's safer to use the latter predicate
* $f(w_y)$ and $f(t_{i+y-1})$ is not a mismatch - then the former predicate is false, and the latter is also, because RHS is 0 and LHS is non zero.
 
To be consistent with the paper, I'd rather leave version with $A_0(i) = \dots$